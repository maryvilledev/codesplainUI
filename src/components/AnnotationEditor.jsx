import React, { PropTypes } from 'react';
import { Tab, Tabs } from 'material-ui/Tabs';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';

import MarkdownDisplayer from './MarkdownDisplayer';
import markdownLogo from '../../res/markdown-logo.svg';

const styles = {
  bottomContainer: {
    display: 'flex',
    flexFlow: 'row wrap',
    alignItems: 'center',
    marginTop: '0.5rem',
  },
  button: {
    flexFlow: 1,
    flexShrink: 1,
    flexBasis: 'auto',
  },
  cancelButton: {
    flexFlow: 1,
    flexShrink: 1,
    flexBasis: 'auto',
    marginRight: '0.2rem',
  },
  hintText: {
    top: '1rem',
  },
  markdownHintText: {
    color: '#d3d3d3',
  },
  markdownIndicator: {
    paddingLeft: '1rem',
    flexFlow: 1,
    flexShrink: 1,
    flexBasis: 'auto',
  },
  markdownLogo: {
    width: '2rem',
    height: '1rem',
  },
};

class AnnotationEditor extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      annotation: this.props.annotation,
    };
    this.clearAnnotation = this.clearAnnotation.bind(this);
    this.onAnnotationChange = this.onAnnotationChange.bind(this);
    this.saveAnnotation = this.saveAnnotation.bind(this);
  }

  onAnnotationChange(ev, annotation) {
    this.setState({
      annotation,
    });
  }

  clearAnnotation() {
    this.onAnnotationChange(null, '');
    this.props.closeAnnotation();
  }

  saveAnnotation() {
    const { annotation } = this.state;
    this.props.saveAnnotation(annotation);
  }

  render() {
    const { annotation } = this.state;
    return (
      <div>
        <Tabs>
          <Tab
            label="Write"
            onActive={() => { this.textField.focus(); }}
          >
            <TextField
              autoFocus
              fullWidth
              hintStyle={styles.hintText}
              hintText="Enter your annotation here"
              multiLine
              name="annotationEditor"
              onChange={this.onAnnotationChange}
              ref={(textField) => { this.textField = textField; }}
              rows={4}
              value={annotation}
            />
          </Tab>
          <Tab label="Preview">
            <MarkdownDisplayer annotation={annotation} />
          </Tab>
        </Tabs>
        <div style={styles.bottomContainer}>
          <RaisedButton
            label="Cancel"
            onTouchTap={this.clearAnnotation}
            secondary
            style={styles.cancelButton}
          />
          <RaisedButton
            disabled={!annotation}
            label="Save"
            onTouchTap={this.saveAnnotation}
            primary
            style={styles.button}
          />
          <a
            href="http://commonmark.org/help"
            rel="noopener noreferrer"
            style={styles.markdownIndicator}
            target="_blank"
          >
            <img
              alt="Markdown Logo"
              src={markdownLogo}
              style={styles.markdownLogo}
            />
            <span style={styles.markdownHintText}>
              Styling with Markdown is supported
            </span>
          </a>
        </div>
      </div>
    );
  }
}

AnnotationEditor.defaultProps = {
  annotation: '',
};

AnnotationEditor.propTypes = {
  annotation: PropTypes.string,
  closeAnnotation: PropTypes.func.isRequired,
  saveAnnotation: PropTypes.func.isRequired,
};

export default AnnotationEditor;
