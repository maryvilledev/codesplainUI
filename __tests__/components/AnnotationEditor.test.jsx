import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

import AnnotationEditor from '../../src/components/AnnotationEditor';

const mockFunctionProps = {
  closeAnnotation: jest.fn(),
  saveAnnotation: jest.fn(),
};

describe('<AnnotationEditor />', () => {
  const muiTheme = getMuiTheme();
  const shallowWithContext = node => shallow(node, { context: { muiTheme } });

  it('matches snapshot', () => {
    const wrapper = shallowWithContext(
      <AnnotationEditor
        {...mockFunctionProps}
      />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('prop: annotation', () => {
    it('renders an empty Textfield if empty', () => {
      const wrapper = shallowWithContext(
        <AnnotationEditor
          {...mockFunctionProps}
        />,
      );
      expect(wrapper.find('TextField').prop('value')).toEqual('');
    });
    it('renders its contents into the TextField', () => {
      const annotation = 'Wubba lubba dub dub!';
      const wrapper = shallowWithContext(
        <AnnotationEditor
          annotation={annotation}
          {...mockFunctionProps}
        />,
      );
      expect(wrapper.find('TextField').prop('value')).toEqual(annotation);
    });
  });

  describe('save button', () => {
    it('is disabled when the textfield is empty', () => {
      const wrapper = shallowWithContext(
        <AnnotationEditor
          {...mockFunctionProps}
        />,
      );
      expect(wrapper.find('[label="Save"]').prop('disabled')).toBe(true);
    });
    it('is enabled when the textfield is not empty', () => {
      const annotation = 'Wubba lubba dub dub!';
      const wrapper = shallowWithContext(
        <AnnotationEditor
          annotation={annotation}
          {...mockFunctionProps}
        />,
      );
      expect(wrapper.find('[label="Save"]').prop('disabled')).toBe(false);
    });
  });
});
