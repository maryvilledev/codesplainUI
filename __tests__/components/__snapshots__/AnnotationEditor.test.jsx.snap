exports[`<AnnotationEditor /> matches snapshot 1`] = `
<div>
  <Tabs
    initialSelectedIndex={0}
    onChange={[Function]}>
    <Tab
      label="Write"
      onActive={[Function]}>
      <TextField
        autoFocus={true}
        disabled={false}
        floatingLabelFixed={false}
        floatingLabelText="Annotation"
        fullWidth={true}
        hintText="Enter your annotation here"
        multiLine={true}
        name="annotationEditor"
        onChange={[Function]}
        rows={4}
        type="text"
        underlineShow={true}
        value="" />
    </Tab>
    <Tab
      label="Preview">
      <MarkdownRenderer
        markdown=""
        options={
          Object {
            "highlight": [Function],
            "langPrefix": "hljs language-",
          }
        } />
    </Tab>
  </Tabs>
  <RaisedButton
    disabled={false}
    fullWidth={false}
    label="Cancel"
    labelPosition="after"
    onTouchTap={[Function]}
    primary={false}
    secondary={true} />
  <RaisedButton
    disabled={true}
    fullWidth={false}
    label="Save"
    labelPosition="after"
    onTouchTap={[Function]}
    primary={true}
    secondary={false} />
</div>
`;
